File: apps/ios/Barns/BarnsApp/Presentation/Care/CareTaskDetailViewModel.swift
import Combine

@MainActor
final class CareTaskDetailViewModel: ObservableObject {
    enum State: Equatable {
        case loading
        case loaded(CareTask)
        case notFound
        case failed(String)
    }

    @Published private(set) var state: State = .loading
    @Published private(set) var isCompleting = false

    private let taskId: CareTask.ID
    private let getCareTaskDetailUseCase: GetCareTaskDetailUseCase
    private let completeCareTaskUseCase: CompleteCareTaskUseCase

    init(
        taskId: CareTask.ID,
        getCareTaskDetailUseCase: GetCareTaskDetailUseCase,
        completeCareTaskUseCase: CompleteCareTaskUseCase
    ) {
        self.taskId = taskId
        self.getCareTaskDetailUseCase = getCareTaskDetailUseCase
        self.completeCareTaskUseCase = completeCareTaskUseCase
    }

    func load() async {
        state = .loading
        do {
            if let task = try await getCareTaskDetailUseCase.execute(id: taskId) {
                state = .loaded(task)
            } else {
                state = .notFound
            }
        } catch {
            state = .failed("Unable to load this task. Please try again.")
        }
    }

    func complete() async {
        guard case .loaded(let task) = state, task.status != .completed, !isCompleting else { return }
        isCompleting = true
        defer { isCompleting = false }
        do {
            try await completeCareTaskUseCase.execute(id: task.id)
            await load()
        } catch {
            state = .failed("Unable to complete this task. Please try again.")
        }
    }
}

File: apps/android/app/src/main/java/com/barns/app/presentation/care/CareTaskDetailViewModel.kt
package com.barns.app.presentation.care

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.barns.app.domain.model.CareTask
import com.barns.app.domain.model.CareTaskStatus
import com.barns.app.domain.usecase.care.CompleteCareTaskUseCase
import com.barns.app.domain.usecase.care.GetCareTaskDetailUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class CareTaskDetailViewModel(
    private val taskId: String,
    private val getCareTaskDetailUseCase: GetCareTaskDetailUseCase,
    private val completeCareTaskUseCase: CompleteCareTaskUseCase,
) : ViewModel() {
    sealed interface State {
        object Loading : State
        data class Loaded(val task: CareTask) : State
        object NotFound : State
        data class Failed(val message: String) : State
    }

    private val _state = MutableStateFlow<State>(State.Loading)
    val state: StateFlow<State> = _state.asStateFlow()

    private val _isCompleting = MutableStateFlow(false)
    val isCompleting: StateFlow<Boolean> = _isCompleting.asStateFlow()

    fun load() {
        _state.value = State.Loading
        viewModelScope.launch {
            runCatching { getCareTaskDetailUseCase.execute(taskId) }
                .onSuccess { task ->
                    _state.value = if (task != null) State.Loaded(task) else State.NotFound
                }
                .onFailure { _state.value = State.Failed("Unable to load this task. Please try again.") }
        }
    }

    fun complete() {
        val current = _state.value
        if (current !is State.Loaded) return
        if (current.task.status == CareTaskStatus.COMPLETED || _isCompleting.value) return
        _isCompleting.value = true
        viewModelScope.launch {
            runCatching {
                completeCareTaskUseCase.execute(current.task.id)
                getCareTaskDetailUseCase.execute(taskId)
            }
                .onSuccess { task ->
                    _state.value = if (task != null) State.Loaded(task) else State.NotFound
                }
                .onFailure { _state.value = State.Failed("Unable to complete this task. Please try again.") }
            _isCompleting.value = false
        }
    }
}
